src/C-DEngine/C-DBaseClasses/cde_KPIs.cs

```diff
@@ -1,604 +1,738 @@
+            nameof(eKPINames.TotalEngineErrors), nameof(eKPINames.TotalEventTimeouts), nameof(eKPINames.UnsignedNodes), nameof(eKPINames.UnsignedPlugins),
+            nameof(eKPINames.UniqueMeshes), nameof(eKPINames.SeenBeforeCount), nameof(eKPINames.KnownNMINodes), nameof(eKPINames.StreamsNotFound),
+            nameof(eKPINames.BlobsNotFound), nameof(eKPINames.WSTestClients), nameof(eKPINames.QSenders), nameof(eKPINames.QSenderInRegistry),
+            nameof(eKPINames.SessionCount), nameof(eKPINames.KPI3)
+        };
+
+
+        // Array of KPI names that should not have a total computed when harvesting to TheThing
+        private static readonly List<string> doNotComputeTotals = new ()
+        {
+            nameof(eKPINames.UnsignedNodes), nameof(eKPINames.UnsignedPlugins), nameof(eKPINames.UniqueMeshes), nameof(eKPINames.BruteDelay),
+            nameof(eKPINames.QSenders), nameof(eKPINames.QSenderInRegistry), nameof(eKPINames.SetPsFired), nameof(eKPINames.TotalEngineErrors),
+            nameof(eKPINames.EngineErrors), nameof(eKPINames.EventTimeouts), nameof(eKPINames.TotalEventTimeouts), nameof(eKPINames.QSNotRelayed),
+            nameof(eKPINames.SessionCount), nameof(eKPINames.WSTestClients), nameof(eKPINames.KPI3), nameof(eKPINames.KPI7), nameof(eKPINames.KPI8), nameof(eKPINames.KPI9)
+        };
+
+        internal static Dictionary<string, object> GetDictionary()
+        {
+            TheThing tn = new ();
+            ToThingProperties(tn, !TheCommonUtils.CBool(TheBaseAssets.MySettings.GetSetting("EnableKPIs")));
+            return tn.GetPBAsDictionary();
+        }
+
+        /// <summary>
+        /// Resets all the KPIs to zero
+        /// </summary>
+        public static void Reset()
+        {
+            // Intentionally left blank.
+
+            // Reason: Reset did NOT reset all KPIs.
+            // Instead it was initializing the KPIs if not yet done which is not necessary anymore.
+        }
+            
+        /// <summary>
+        /// Returns a string of all KPIs
+        /// </summary>
+        /// <param name="doReset">if true, the KPIs are reset after the string was constructed.</param>
+        /// <returns></returns>
+        public static string GetKPIs(bool doReset)
+        {
+            StringBuilder tRes = new ();
+            tRes.Append($"C-DEngine-KPIs: LR:{LastReset:MM/dd/yyyy hh:mm:ss.fff--tt} ");
+            List<string> orderedKeys = KPIs.Keys.OrderBy(s => s).ToList();
+            foreach (string key in orderedKeys)
+            {
+                tRes.Append($"{key}:{GetKPI(key)} ");
+            }
+            if (doReset) Reset();
+            return tRes.ToString();
+        }
+
+        private static readonly object thingHarvestLock = new ();
+        internal static void ToThingProperties(TheThing pThing, bool bReset)
+        {
+            if (pThing == null) return;
+
+            lock (thingHarvestLock)
+            {
+                _syncKPIs.EnterUpgradeableReadLock();

```

Comment:
This change is great: guarantees that we harvest a consistent snapshot of all KPIs.
However, we are doing some relatively expensive operations here (JSON serialization, Thing Property get/set's). Could we maybe "harvest" the KPI values into a local datastructure first, and then - still under the thingHarvestLock but outside the _syncKPIs lock - do the more expensive processing?
It seems that while the harvest is ongoing, essentially not messages will be sent or processed because we log KPIs.

Reply:
This lock is not new in this PR. While the statement is true, we could do some more research on how much impact this lock has on Cloud performance before we change it.

Reply:
I understand Markus concerns and the statement is definitely true.

I tried find an effortless solution for this by avoiding execution of the method in case it is already running, which mainly affects cyclic KPI harvesting. In any other case, execution will be synchronized by forcable aquiring the lock.
Cyclic execution can also be controlled with the setting `KPIHarvestIntervalInSeconds` in case KPI harvesting takes too long or increases CPU load.

Also the amount of KPIs is not that much at the moment. So maybe the proposed new change in lock-handling is enough for now?

Reply:
My concern is that all message processing (and any other operation that reports KPis) halts while we harvest, due to the read lock on all KPIs. We should try to do the absolute minimum work while holding this global lock and also not call out to any external code (SetProperty can be intercepted by plugins). The harvest lock is fine, because it only blocks other harvesters and they are not time sensitive.

Here's what I have in mind (compiles but haven't tested): https://github.com/TRUMPF-IoT/C-DEngine/pull/66/files.

Reply:
Ah - now I fully understand your concerns.
I merged your changes into this branch and added an optimization which resets the KPIs during the harvesting them into the local data structure.

Reply:
Perfect!